File: src/mVasu.Pwa/src/app/core/models/task.dto.ts
/**
 * Task queue DTOs — mirror the contracts described in
 * design/handoff_navigation/design_handoff_mvasu_navigation/BACKEND.md.
 *
 * Backend wiring is not yet in place; the PWA renders the Task tab from a
 * mock fixture today. When the API is ready, these types stay; only the
 * service that produces them changes.
 */

export type TaskType =
  | 'visit-introduction'
  | 'visit-reservation'
  | 'open-house'
  | 'signature-pending'
  | 'inbox-termination'
  | 'inbox-signed'
  | 'photo-scheduled'
  | 'renovation-approval'
  | 'lead-callback'
  | 'desk-list-item';

export type TaskAccent = 'navy' | 'cta' | 'info' | 'warn';

export type TaskActionKind =
  | 'navigate'
  | 'phone'
  | 'sms'
  | 'mark-done'
  | 'cancel'
  | 'external';

export interface TaskAction {
  readonly kind: TaskActionKind;
  readonly label: string;
  readonly href?: string;
  readonly destructive?: boolean;
  readonly primary?: boolean;
}

export interface TaskWhen {
  readonly time: string;
  readonly note?: string;
}

export interface TaskEntityRef {
  readonly module: string;
  readonly id: string;
}

export type TaskGroupId = 'today' | 'tomorrow' | 'this-week' | 'later';

export interface TaskCard {
  readonly id: string;
  readonly type: TaskType;
  readonly accent: TaskAccent;
  readonly urgent: boolean;
  readonly when: TaskWhen;
  readonly title: string;
  readonly who?: string;
  readonly meta?: string;
  readonly actions: ReadonlyArray<TaskAction>;
  readonly entityRef: TaskEntityRef;
  /** Optional explicit sort hint; lower comes first inside a group. */
  readonly sortOrder?: number;
  /**
   * Optional caption override for the card's type-row label. When present,
   * overrides the static TaskType label — used for `xVasuSecuritySystemUserTask`
   * cards where `UserTaskType.Name` (e.g. "Hinnoittelu", "Tarkastus") drives
   * the caption instead of the accent-bucket `type`.
   */
  readonly typeLabel?: string;
}

export interface TaskGroup {
  readonly id: TaskGroupId;
  readonly label: string;
  readonly date: string;
  readonly tasks: ReadonlyArray<TaskCard>;
}

export interface TasksState {
  readonly groups: ReadonlyArray<TaskGroup>;
  readonly loading: boolean;
  readonly refreshing: boolean;
  readonly error: string | null;
  readonly lastFetched: Date | null;
}

File: src/mVasu.Api.Contracts/TaskDto.cs
namespace mVasu.Api.Contracts;

/// <summary>
/// Heterogeneous task surfaced in the Tehtävät tab. Aggregates rows from
/// multiple XAF entity sources (xVasuSecuritySystemUserTask, Tutustumis-
/// kaynti, Yleisesittely, Varausesittely, Tarjous, Sopimus, Saapuneet
/// irtisanomiset, Valokuvaus, Remontti, Liidi, Tiskilista) into a single
/// stream. The mapping rules live in BACKEND.md §2 and EXISTING_ENTITIES.md.
/// </summary>
/// <remarks>
/// <para>The frontend consumes <c>Type</c> / <c>Accent</c> as strings to
/// keep the JSON wire format stable across XAF schema additions; the
/// canonical set lives in <see cref="TaskTypeNames"/> and
/// <see cref="TaskAccents"/>.</para>
/// </remarks>
public sealed record TaskDto(
    string Id,
    /// <summary>One of the values in <see cref="TaskTypeNames"/>.</summary>
    string Type,
    /// <summary>One of the values in <see cref="TaskAccents"/>.</summary>
    string Accent,
    bool Urgent,
    TaskWhenDto When,
    string Title,
    string? Who,
    string? Meta,
    /// <summary>
    /// Optional caption override for the row's type label. Used for
    /// xVasuSecuritySystemUserTask rows where <c>UserTaskType.Name</c>
    /// (e.g. "Hinnoittelu", "Tarkastus") drives the row caption instead
    /// of the static <see cref="Type"/> mapping.
    /// </summary>
    string? TypeLabel,
    IReadOnlyList<TaskActionDto> Actions,
    TaskEntityRefDto EntityRef,
    int? SortOrder);

/// <summary>"When" block: human time + optional secondary note + sort hint.</summary>
public sealed record TaskWhenDto(
    string Time,
    string? Note,
    /// <summary>
    /// Unix-millis timestamp for chronological sort, or 0 for "Heti"
    /// (out-of-band / overdue). Used only on the server / for sorting.
    /// </summary>
    long SortOrder);

/// <summary>
/// Back-reference to the originating XAF entity. Module values track
/// NAVIGATION.md §2 ("asma" / "kire" / "verkkokauppa" / "core").
/// </summary>
public sealed record TaskEntityRefDto(
    string Module,
    string EntityType,
    string Id);

/// <summary>Canonical task-type identifiers — see BACKEND.md §1.</summary>
public static class TaskTypeNames
{
    public const string VisitIntroduction  = "visit-introduction";   // Tutustumiskaynti
    public const string VisitReservation   = "visit-reservation";    // Varausesittely
    public const string OpenHouse          = "open-house";           // Yleisesittely
    public const string SignaturePending   = "signature-pending";    // Tarjous / Sopimus
    public const string InboxTermination   = "inbox-termination";    // Saapunut irtisanomisilmoitus
    public const string InboxSigned        = "inbox-signed";         // Palautunut sähköinen allekirjoitus
    public const string PhotoScheduled     = "photo-scheduled";      // Valokuvaus
    public const string RenovationApproval = "renovation-approval";  // Remontti
    public const string LeadCallback       = "lead-callback";        // Liidi
    public const string DeskListItem       = "desk-list-item";       // Tiskilistan rivi
    public const string GenericTask        = "generic-task";         // xVasuSecuritySystemUserTask
}

/// <summary>Card accent (border-left + primary CTA tint) — see SCREENS.md §01.</summary>
public static class TaskAccents
{
    public const string Navy = "navy";
    public const string Cta  = "cta";
    public const string Info = "info";
    public const string Warn = "warn";
}
